src/authenticator.rs

```diff
@@ -1001,6 +1044,114 @@ where
         )
     }
 
+    fn _extension_pin_factory_reset(&mut self) -> Result {
+        try_syscall!(self.trussed.delete_pin(BACKEND_USER_PIN_ID))
+            .map_err(|_| iso7816::Status::UnspecifiedNonpersistentExecutionError)?;
+        Ok(())
+    }
+
+    fn _extension_check_pin(&mut self, password: &[u8]) -> Result {
+        let reply = try_syscall!(self.trussed.check_pin(
+            BACKEND_USER_PIN_ID,
+            Bytes::from_slice(password)
+                .map_err(|_| iso7816::Status::UnspecifiedNonpersistentExecutionError)?
+        ))
+        .map_err(|_| iso7816::Status::UnspecifiedNonpersistentExecutionError)?;
+        if !(reply.success) {
+            Err(Status::SecurityStatusNotSatisfied)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn _extension_set_pin(&mut self, password: &[u8]) -> Result {
+        try_syscall!(self.trussed.set_pin(
+            BACKEND_USER_PIN_ID,
+            Bytes::from_slice(password)
+                .map_err(|_| iso7816::Status::UnspecifiedNonpersistentExecutionError)?,
+            Some(ATTEMPT_COUNTER_DEFAULT_RETRIES),
+            true
+        ))
+        .map_err(|_| iso7816::Status::UnspecifiedNonpersistentExecutionError)?;
+        Ok(())
+    }
+
+    fn _extension_change_pin<'l>(&mut self, password: &'l [u8], new_password: &'l [u8]) -> Result {
+        self._extension_check_pin(password)?;
+        self._extension_set_pin(new_password)?;
```

Comment:
I think this should use `AuthClient::change_pin` directly so that the PIN key is not lost.  And maybe also add a unit test for that?

Reply:
Good suggestion

Reply:
Implemented in #41 and added test for data retention:
- https://github.com/Nitrokey/pynitrokey/pull/329/files#diff-8fb21877f9a0cc6c88d7d4076e5e6ab211a2bb69dd9d7cfa795c3297e1f4e9e7R780